File: 056.MergeIntervals.cpp
class Solution {
public:
    static bool lesser(const Interval& a, const Interval& b) {
        if(a.start == b.start)
            return a.end < b.end;
        return a.start < b.start;
    }
    vector<Interval> merge(vector<Interval> &intervals) {
        if (intervals.size() <= 1) {
            return intervals;
        }
        sort(all(intervals), lesser);
        vector<Interval> result;
        Interval v = intervals[0];
        int i = 1;
        while (i < intervals.size()) {
            if (v.end >= intervals[i].start) {
                if (intervals[i].end > v.end) {
                    v.end = intervals[i].end;
                }
            } else {
                result.push_back(v);
                v = intervals[i];
            }
            i ++;
        }
        result.push_back(v);
        return result;
    }
};

File: Algorithms/py/056.MergeIntervals.py
class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[Interval]
        """
        if len(intervals) <= 1:
            return intervals
        intervals.sort(key=lambda x: x[0])
        # TODO
        # the following is wrong
        #intervals = sorted(intervals)
        result = []
        v = intervals[0]
        i = 1
        while i < len(intervals):
            if v[1] >= intervals[i][0]:
                if intervals[i][1] > v[1]:
                    v[1] = intervals[i][1];
            else:
                result.append(v)
                v = intervals[i]
            i += 1
        result.append(v)
        return result